solver/src/driver.rs

```diff
@@ -104,91 +105,56 @@ impl Driver {
-            liquidity
-                .into_iter()
-                .partition(|liquidity| match liquidity {
-                    Liquidity::Limit(limit_order) => {
-                        [limit_order.sell_token, limit_order.buy_token]
-                            .iter()
-                            .all(|token| estimated_prices.contains_key(token))
-                    }
-                    Liquidity::Amm(_) => true,
-                });
-        if !removed_orders.is_empty() {
-            tracing::debug!(
-                "pruned {} orders: {:?}",
-                removed_orders.len(),
-                removed_orders,
-            );
-        }
-
-        self.metrics.liquidity_fetched(&liquidity);
-
-        let gas_price = self
-            .gas_price_estimator
-            .estimate()
-            .await
-            .context("failed to estimate gas price")?;
-        tracing::debug!("solving with gas price of {}", gas_price);
-
-        let mut settlements: Vec<(String, Settlement)> =
-            join_all(self.solver.iter().map(|solver| {
-                let liquidity = liquidity.clone();
-                let metrics = &self.metrics;
-                async move {
-                    let label = format!("{}", solver);
-                    let start_time = Instant::now();
-                    let settlement = solver.solve(liquidity, gas_price).await;
-                    metrics.settlement_computed(&label, start_time);
-                    (label, settlement)
-                }
-            }))
-            .await
-            .into_iter()
-            .filter_map(|(solver, settlement)| match settlement {
-                Ok(settlement) => settlement.map(|settlement| (solver, settlement)),
+    async fn run_solvers(
+        &self,
+        liquidity: Vec<Liquidity>,
+        gas_price: f64,
+        prices: &HashMap<H160, BigRational>,
+    ) -> Vec<SolverSettlement> {
+        join_all(self.solver.iter().enumerate().map(|(index, solver)| {
+            let liquidity = liquidity.clone();
+            let metrics = &self.metrics;
+            async move {
+                let start_time = Instant::now();
+                let settlement = solver.solve(liquidity, gas_price).await;
+                metrics.settlement_computed(solver.to_string().as_str(), start_time);
+                (index, settlement)
+            }
+        }))
+        .await
+        .into_iter()
+        .filter_map(|(index, settlement)| {
+            let settlement = match settlement {
+                Ok(settlement) => settlement,
                 Err(err) => {
-                    tracing::error!("solver {} error: {:?}", solver, err);
-                    None
+                    tracing::error!("solver {} error: {:?}", self.solver[index].to_string(), err);
+                    return None;
                 }
+            }?;
+            if settlement.trades().is_empty() {
+                return None;
+            }
```

Comment:
This is a small logic change because we were previously doing this in the settlement transaction submission loop below and printing "skipping empty settlement". I figured this makes more sense as we are also filtering `None` out here and there is no reason to log this happening.